### PR TITLE
[#3] [CHORE] Setup a new Github Action workflow to verify the code changes in pull request

### LINKED ISSUE #3: [CHORE] Setup a new Github Action workflow to verify the code changes in pull request
## Why
We have to set up a new CI workflow to ensure the new code work properly and they don't cause any trouble.

## Acceptance Criteria
- Create a new `test` workflow. This workflow has some steps:
  - Checkout code.
  - Install the necessary tools for React JS
  - Do lint check
  - Run unittest

### PR DESCRIPTION
**Pull request by manh-t**
Close #3 

## What happened 👀

- Add the `test` workflow to verify the code changes in a pull request.
- Migrate `cypress` in order to launch the integration test. [Migration guide](https://docs.cypress.io/guides/references/migration-guide)

## Insight 📝

- I create a `PERSONAL_GITHUB_TOKEN` secret to apply `danger`

## Proof Of Work 📹

- The `test` workflow should pass in this PR.
- Danger setup properly.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -1,42 +1,42 @@
...
 
-#       - name: Setup node and restore cached dependencies
-#         uses: actions/setup-node@v3
-#         with:
-#           node-version: '16'
-#           cache: 'npm'
+      - name: Setup node and restore cached dependencies
+        uses: actions/setup-node@v3
+        with:
+          node-version: '16'
```
Comment (nvminhtue) [inline .github/workflows/test.yml]: Regarding [this comment](https://github.com/manh-t/react-survey/pull/39#discussion_r1222909481), let's update this one as well

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,42 +1,42 @@
...
 
-#       - name: Run linter
-#         run: npm run codebase:lint
+      - name: Run linter
+        run: npm run codebase:lint
 
-#       - name: Run unit tests
-#         run: npm test -- --coverage
+      - name: Run unit tests
+        run: npm test -- --coverage
```
Comment (nvminhtue) [inline .github/workflows/test.yml]: isn't it?
```suggestion
        run: npm test --coverage
```

File: .github/workflows/test.yml (hunk 1 above)
Comment (manh-t) [reply .github/workflows/test.yml]: @nvminhtue updated in [1c51466](https://github.com/manh-t/react-survey/commit/1c514664e38d1087e281b805600b731bf43c76c3)

File: .github/workflows/test.yml (hunk 2 above)
Comment (manh-t) [reply .github/workflows/test.yml]: @nvminhtue it seems we have to add an `--` to add extra parameters to the script of npm. Otherwise it won't work 🙏

**Comment by manh-t** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Congrats, coverage is good</td>
    </tr>
  </tbody>
</table>

## Coverage in All Files
File | Line Coverage | Statement Coverage | Function Coverage | Branch Coverage
---- | ------------: | -----------------: | ----------------: | --------------:
[src/App.tsx](../blob/0fea9f0/src/App.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/index.tsx](../blob/0fea9f0/src/index.tsx) | (3/3) 100% | (3/3) 100% | (0/0) NaN% | (0/0) NaN%
[src/lib/requestManager.ts](../blob/0fea9f0/src/lib/requestManager.ts) | (5/5) 100% | (5/5) 100% | (2/2) 100% | (1/1) 100%
[src/routes/index.tsx](../blob/0fea9f0/src/routes/index.tsx) | (1/1) 100% | (1/1) 100% | (0/0) NaN% | (0/0) NaN%
[src/screens/Home/index.tsx](../blob/0fea9f0/src/screens/Home/index.tsx) | (2/2) 100% | (2/2) 100% | (1/1) 100% | (0/0) NaN%
Total | (14/14) 100% | (14/14) 100% | (4/4) 100% | (1/1) 100%

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d96246a54f037a74bfa29010cbfb1dfbb60c6d73
</p>

File: package.json [hunk 3]
```diff
@@ -15,7 +15,7 @@
     "build": "react-scripts build",
     "test": "react-scripts test --transformIgnorePatterns",
     "eject": "react-scripts eject",
-    "test:coverage": "rimraf coverage && react-scripts test --coverage --watchAll=false && npm run cypress:run && node ./scripts/coverage-merge.js && nyc report",
+    "test:coverage": "rimraf coverage && npm test -- --coverage --watchAll=false && npm run cypress && npm run test:merge-coverage && nyc report",
```
Comment (nvminhtue) [inline package.json]: it's just a small suggestion, you can take a look on [concurrently](https://www.npmjs.com/package/concurrently) package to help the script shorter

Besides, I don't think we have to remove the `coverage` folder at the beginning

**Review by nvminhtue — APPROVED** [review]
LGTM 👍

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+jobs:
+  test:
+    name: Run linters and tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Cancel Previous Runs
+        uses: styfle/cancel-workflow-action@0.10.0
+        with:
+          access_token: ${{ github.token }}
```
Comment (tyrro) [inline .github/workflows/test.yml]: GitHub actions support native [concurrency](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency), could we use that?

```bash
# before Line 5
concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true
```

File: package.json (hunk 3 above)
Comment (manh-t) [reply package.json]: @nvminhtue nice, I will have a try on `concurrently` package 👍 Thanks for the suggestion 😄 
`rimraf coverage` was in the template, let me remove it

**Review by tyrro — COMMENTED** [review]
could you take care of this warning as well?

image

File: .github/workflows/test.yml (hunk 4 above)
Comment (manh-t) [reply .github/workflows/test.yml]: cool, let me update it

**Comment by manh-t** [comment]
> could you take care of this warning as well?
> 
> image

@tyrro I create a chore #45  to handle this issue later 🙏

**Review by tyrro — APPROVED** [review]
well done!  🚀

**Pull request merged**